core/src/main/java/org/apache/iceberg/view/BaseViewOperations.java

```diff
@@ -0,0 +1,212 @@
+    doCommit(base, metadata);
+    requestRefresh();
+
+    LOG.info(
+        "Successfully committed to view {} in {} ms",
+        viewName(),
+        System.currentTimeMillis() - start);
+  }
+
+  protected String writeNewMetadata(ViewMetadata metadata, int newVersion) {
+    String newMetadataFilePath = newMetadataFilePath(metadata, newVersion);
+    OutputFile newMetadataLocation = io().newOutputFile(newMetadataFilePath);
+
+    // write the new metadata
+    // use overwrite to avoid negative caching in S3. this is safe because the metadata location is
+    // always unique because it includes a UUID.
+    ViewMetadataParser.overwrite(metadata, newMetadataLocation);
+
+    return newMetadataLocation.location();
+  }
+
+  private String newMetadataFilePath(ViewMetadata metadata, int newVersion) {
+    return metadataFileLocation(
+        metadata, String.format("%05d-%s%s", newVersion, UUID.randomUUID(), ".metadata.json"));
+  }
+
+  private String metadataFileLocation(ViewMetadata metadata, String filename) {
+    return String.format("%s/%s/%s", metadata.location(), METADATA_FOLDER_NAME, filename);
+  }
+
+  protected void refreshFromMetadataLocation(String newLocation) {
+    refreshFromMetadataLocation(newLocation, null, 20);
+  }
+
+  protected void refreshFromMetadataLocation(
+      String newLocation, Predicate<Exception> shouldRetry, int numRetries) {
+    refreshFromMetadataLocation(
+        newLocation,
+        shouldRetry,
+        numRetries,
+        metadataLocation -> ViewMetadataParser.read(io().newInputFile(metadataLocation)));
+  }
+
+  protected void refreshFromMetadataLocation(
+      String newLocation,
+      Predicate<Exception> shouldRetry,
+      int numRetries,
+      Function<String, ViewMetadata> metadataLoader) {
+    if (!Objects.equal(currentMetadataLocation, newLocation)) {
+      LOG.info("Refreshing view metadata from new version: {}", newLocation);
+
+      AtomicReference<ViewMetadata> newMetadata = new AtomicReference<>();
+      Tasks.foreach(newLocation)
+          .retry(numRetries)
+          .exponentialBackoff(100, 5000, 600000, 4.0 /* 100, 400, 1600, ... */)
+          .throwFailureWhenFinished()
+          .stopRetryOn(NotFoundException.class) // overridden if shouldRetry is non-null
+          .shouldRetryTest(shouldRetry)
+          .run(metadataLocation -> newMetadata.set(metadataLoader.apply(metadataLocation)));
+
+      this.currentMetadata = newMetadata.get();
+      this.currentMetadataLocation = newLocation;
+      this.version = parseVersion(newLocation);
+    }
+
+    this.shouldRefresh = false;
+  }
+
+  /**
+   * Parse the version from view metadata file name.
+   *
+   * @param metadataLocation view metadata file location
+   * @return version of the view metadata file in success case and -1 if the version is not parsable
+   *     (as a sign that the metadata is not part of this catalog)
+   */
+  private static int parseVersion(String metadataLocation) {
```

Comment:
there are a few things in this class that are code duplicates of `BaseMetastoreTableOperations` so we might want to consider refactoring this out and make it re-usable

Reply:
This changes so seldom that I think we're okay for now.